Mixed integer optimization

hongkai-dai opened the pull request:
I tested simple integer linear program with Gurobi 6.5 and Mosek.

I will create a PR later to support mixed-integer linear program, MIQP, MISOCP etc. Also Gurobi 7 supports indicator constraint, @rdeits has done great experiment on this new feature already. We can consider to support it in the future.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/3948)
<!-- Reviewable:end -->

hongkai-dai commented:
+@rdeits for feature review. +@russtedrake if you are interested.

---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-:-KVC3uOJA_cPFUMCaOhH:b4izgrx)*
<!-- Sent from Reviewable.io -->

rdeits commented:
Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/solvers/test/mathematical_program_test_util.h, line 23 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCAuJ1KstZm994LS0b:-KVCAuJ25pXvAwwKb66E:b-sdulp2) ([raw file](https://github.com/robotlocomotion/drake/blob/c0a927ad95b21483bddc3fb22807188b2e867654/drake/solvers/test/mathematical_program_test_util.h#L23)):*
> ```objc
> }
> 
> void AddSolverToList(
> ```

I would suggest changing the name of this function. Right now, the name suggests that it will always add the solver to the given list, while it will actually only do so if the solver is available. And I think the "ToList" part of the name is redundant given that the argument type is already a std::list. Perhaps "AddSolverIfAvailable" or similar?

---

*[drake/solvers/test/mixed_integer_optimization_test.cc, line 42 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCBwayRBPgt1mM4H8r:-KVCBwayRBPgt1mM4H8s:bc93g93) ([raw file](https://github.com/robotlocomotion/drake/blob/c0a927ad95b21483bddc3fb22807188b2e867654/drake/solvers/test/mixed_integer_optimization_test.cc#L42)):*
> ```cpp
>     prog.AddLinearConstraint(
>         a2, drake::Vector1d(1),
>         drake::Vector1d(std::numeric_limits<double>::infinity()), {x(0), x(1)});
> ```

This is outside the domain of this pull request, but I think that requiring 1-element vectors for scalar bounds is going to be annoying. Would it be possible to add another AddLinearConstraint method that takes a row-vector of coefficients and scalar upper and lower bounds?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3948)*
<!-- Sent from Reviewable.io -->

rdeits commented:
This looks great. For prototyping, I would still want to interact at a higher level of abstraction (i.e. building up symbolic expressions and constraints rather than specifying coefficients), but this seems like the right lower-level interface. 

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-:-KVCCYo3QM2C51ZOfyal:b-ya8a50)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Thank @rdeits for the review. I made some changes, PTAL. 

I very much agree on the symbolic expression idea. @soonho-tri is currently working on that. And we have a github issue as the roadmap https://github.com/RobotLocomotion/drake/issues/3906.

---

Review status: 4 of 9 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/solvers/test/mathematical_program_test_util.h, line 23 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCAuJ1KstZm994LS0b:-KVCcV0Vv7QB1_R3B7xl:b1axc77) ([raw file](https://github.com/robotlocomotion/drake/blob/c0a927ad95b21483bddc3fb22807188b2e867654/drake/solvers/test/mathematical_program_test_util.h#L23)):*
<details><summary><i>Previously, rdeits (Robin Deits) wrote…</i></summary>
> I would suggest changing the name of this function. Right now, the name suggests that it will always add the solver to the given list, while it will actually only do so if the solver is available. And I think the "ToList" part of the name is redundant given that the argument type is already a std::list. Perhaps "AddSolverIfAvailable" or similar?
</details>
Done. Good Point, I changed the name to `AddSolverToListIfAvailable`, I still want to emphasize that you are using a list of solvers, not a other containers such as vector.

---

*[drake/solvers/test/mixed_integer_optimization_test.cc, line 42 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCBwayRBPgt1mM4H8r:-KVCcjm5S8-5xdCaUopx:b-qdj9p8) ([raw file](https://github.com/robotlocomotion/drake/blob/c0a927ad95b21483bddc3fb22807188b2e867654/drake/solvers/test/mixed_integer_optimization_test.cc#L42)):*
<details><summary><i>Previously, rdeits (Robin Deits) wrote…</i></summary>
> This is outside the domain of this pull request, but I think that requiring 1-element vectors for scalar bounds is going to be annoying. Would it be possible to add another AddLinearConstraint method that takes a row-vector of coefficients and scalar upper and lower bounds?
</details>
Done. I like the idea. I added following functions to support constraint with only one row
```
AddLinearConstraint
AddLinearEqualityConstraint
AddBoundingBoxConstraint
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-:-KVCd0yrz-E8WN_YpgWA:b-kegdxe)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
+@sherm1 for platform review

---

Review status: 4 of 9 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-:-KVCdO49KY_drURbNY0S:buasbda)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
looks great.  and thanks for doing this so fast!!  I do have one major comment, and a few minor ones.

---

Reviewed 4 of 8 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/solvers/mathematical_program.h, line 702 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCimTwXoGlFgzAW1zy:-KVCimTxLWqu9rLIG6N2:b-r90tcn) ([raw file](https://github.com/robotlocomotion/drake/blob/1b2a663430020208a75e17bb8a05247d5f4c4e9e/drake/solvers/mathematical_program.h#L702)):*
> ```objc
>    * The code above imposes constraints <br>
>    * -x(2) + 2*x(5) = 1 <br>
>    *  x(2) +   x(5) = 3
> ```

mathjax instead?

---

*[drake/solvers/mathematical_program.h, line 1284 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCkChpS88eRKvE2ICt:-KVCkChpS88eRKvE2ICu:b-r7hco5) ([raw file](https://github.com/robotlocomotion/drake/blob/1b2a663430020208a75e17bb8a05247d5f4c4e9e/drake/solvers/mathematical_program.h#L1284)):*
> ```objc
> 
>   // Invariant: variable_type_.length() == num_vars_;
>   std::vector<DecisionVariable::VarType> variable_type_;
> ```

unfortunately, i think the ownership here is wrong.  
the variable already has a type attribute:
https://github.com/RobotLocomotion/drake/blob/master/drake/solvers/decision_variable.h#L22
and I think it's more correct for the variable to own it's type than for the mathematicalprogram to own it.  can we zap this variable and use the other data instead?

---

*[drake/solvers/mosek_solver.cc, line 391 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCjD1TZZ_0lqv6naiB:-KVCj9hH2TVEIyZVh6zc:bb7phvt) ([raw file](https://github.com/robotlocomotion/drake/blob/1b2a663430020208a75e17bb8a05247d5f4c4e9e/drake/solvers/mosek_solver.cc#L391)):*
> ```cpp
>     rescode = AddBoundingBoxConstraints(prog, &task);
>   }
>   // Spcify binary variables.
> ```

typo

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-:-KVCkbvzqOpNYL247Zw9:b-q92g2g)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
This looks very nice. A few minor comments, PTAL.

---

Reviewed 4 of 8 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[drake/solvers/mathematical_program.h, line 381 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCjxedMsBTiApkvuHa:-KVCjxeel9fbFm3sz_1Q:bu2asn3) ([raw file](https://github.com/robotlocomotion/drake/blob/1b2a663430020208a75e17bb8a05247d5f4c4e9e/drake/solvers/mathematical_program.h#L381)):*
> ```objc
>    * and/or constraints to have any effect during optimization.
>    * @param num_new_vars number of new binary variables.
>    * @param name The name of the binary variables, @default is "b".
> ```

BTW: reflow to get rid of nearly-blank line above; capitalize "number"; not sure `@default` is what you want here -- did you generate the doxygen and look at it?

---

*[drake/solvers/mathematical_program.h, line 381 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCkWMx3tTHsasO_0H6:-KVCkWMx3tTHsasO_0H7:bxitblk) ([raw file](https://github.com/robotlocomotion/drake/blob/1b2a663430020208a75e17bb8a05247d5f4c4e9e/drake/solvers/mathematical_program.h#L381)):*
> ```objc
>    * and/or constraints to have any effect during optimization.
>    * @param num_new_vars number of new binary variables.
>    * @param name The name of the binary variables, @default is "b".
> ```

Does it make sense to have a default name here? If I add two binary variables taking the default is it OK that they are both named "b"? Or does it make "b1" and "b2"? Comment should clarify.

---

*[drake/solvers/mathematical_program.h, line 690 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVClv0IM5jW6ahbfGWm:-KVClv0IM5jW6ahbfGWn:b-siwuvn) ([raw file](https://github.com/robotlocomotion/drake/blob/1b2a663430020208a75e17bb8a05247d5f4c4e9e/drake/solvers/mathematical_program.h#L690)):*
> ```objc
>    * @brief Adds linear equality constraints referencing potentially a subset of
>    * the decision variables.
>    * Example: to add and equality constraint which only depends on two of the
> ```

BTW, typo and->an

---

*[drake/solvers/mathematical_program.h, line 692 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVClKLekj9u6pmnIvCk:-KVClKLekj9u6pmnIvCl:b-14ktfe) ([raw file](https://github.com/robotlocomotion/drake/blob/1b2a663430020208a75e17bb8a05247d5f4c4e9e/drake/solvers/mathematical_program.h#L692)):*
> ```objc
>    * Example: to add and equality constraint which only depends on two of the
>    * elements of x, you could use
>    * \code{.cc}
> ```

BTW, use `@code`; we don't allow doxygen backslashes. Also, please verify that doxygen recognizes ".cc" as C++. (Generate and take a look if you haven't already.)

---

*[drake/solvers/mathematical_program.h, line 702 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCm6EZfskKDGLB_4X7:-KVCm6EZfskKDGLB_4X8:b-ry6s6o) ([raw file](https://github.com/robotlocomotion/drake/blob/1b2a663430020208a75e17bb8a05247d5f4c4e9e/drake/solvers/mathematical_program.h#L702)):*
> ```objc
>    * The code above imposes constraints <br>
>    * -x(2) + 2*x(5) = 1 <br>
>    *  x(2) +   x(5) = 3
> ```

Simpler just to use `<pre>` and `</pre>` around these ascii equations to make sure they don't get reformatted. No need for `<br>`.

---

*[drake/solvers/mathematical_program.h, line 728 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCmMFbuWn797SKxzO4:-KVCmMFbuWn797SKxzO5:b4s94hx) ([raw file](https://github.com/robotlocomotion/drake/blob/1b2a663430020208a75e17bb8a05247d5f4c4e9e/drake/solvers/mathematical_program.h#L728)):*
> ```objc
>    * Add one row of linear equality constraint referencing potentially a subset
>    * of decision variables.
>    * a*vars = beq
> ```

BTW, check doxygen output.

---

*[drake/solvers/mathematical_program.h, line 744 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCmRuiMWTB7eASV91q:-KVCmRuiMWTB7eASV91r:b4s94hx) ([raw file](https://github.com/robotlocomotion/drake/blob/1b2a663430020208a75e17bb8a05247d5f4c4e9e/drake/solvers/mathematical_program.h#L744)):*
> ```objc
>    * Add one row of linear equality constraint referencing all
>    * decision variables.
>    * a*vars = beq
> ```

BTW, check doxygen output.

---

*[drake/solvers/mosek_solver.cc, line 334 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCnJ031jh0bllubrZd:-KVCnJ031jh0bllubrZe:b-m10sxn) ([raw file](https://github.com/robotlocomotion/drake/blob/1b2a663430020208a75e17bb8a05247d5f4c4e9e/drake/solvers/mosek_solver.cc#L334)):*
> ```cpp
>       *with_integer_or_binary_variable = true;
>       rescode = MSK_putvartype(*task, i, MSK_VAR_TYPE_INT);
>       double xi_lb, xi_ub;
> ```

BTW, consider initializing these to NaN in case the method below doesn't actually set them for some reason (an error or bug in Mosek?).

---

*[drake/solvers/mosek_solver.cc, line 354 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCngnx4yfEsT2USVLD:-KVCngnx4yfEsT2USVLE:b-t5kh37) ([raw file](https://github.com/robotlocomotion/drake/blob/1b2a663430020208a75e17bb8a05247d5f4c4e9e/drake/solvers/mosek_solver.cc#L354)):*
> ```cpp
>   return rescode;
> }
> }  // end namespace
> ```

BTW, should be "// namespace *whatever*"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-:-KVCoLNMMgj6PA1b_4ug:b-frugzv)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Thank @russtedrake and @sherm1 for the review. I made some changes, PTAL.

---

Review status: 6 of 10 files reviewed at latest revision, 14 unresolved discussions.

---

*[drake/solvers/mathematical_program.h, line 702 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCimTwXoGlFgzAW1zy:-KVCovdAspm-4PHkdtsi:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1b2a663430020208a75e17bb8a05247d5f4c4e9e/drake/solvers/mathematical_program.h#L702)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> mathjax instead?
</details>
Done.

---

*[drake/solvers/mathematical_program.h, line 702 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCm6EZfskKDGLB_4X7:-KVCw_pTXA0HVEQzJHWx:b4cgqny) ([raw file](https://github.com/robotlocomotion/drake/blob/1b2a663430020208a75e17bb8a05247d5f4c4e9e/drake/solvers/mathematical_program.h#L702)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Simpler just to use `<pre>` and `</pre>` around these ascii equations to make sure they don't get reformatted. No need for `<br>`.
</details>
Done. I changed to mathjax.

---

*[drake/solvers/mathematical_program.h, line 728 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCmMFbuWn797SKxzO4:-KVCpMHes9FJlQAclLFE:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1b2a663430020208a75e17bb8a05247d5f4c4e9e/drake/solvers/mathematical_program.h#L728)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, check doxygen output.
</details>
Done.

---

*[drake/solvers/mathematical_program.h, line 744 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCmRuiMWTB7eASV91q:-KVCpN4b_NwwejFu4Gdu:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1b2a663430020208a75e17bb8a05247d5f4c4e9e/drake/solvers/mathematical_program.h#L744)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, check doxygen output.
</details>
Done.

---

*[drake/solvers/mathematical_program.h, line 1284 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCkChpS88eRKvE2ICt:-KVCqi8PdQ5q_OBFOdN7:b-k01xef) ([raw file](https://github.com/robotlocomotion/drake/blob/1b2a663430020208a75e17bb8a05247d5f4c4e9e/drake/solvers/mathematical_program.h#L1284)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> unfortunately, i think the ownership here is wrong.  
> the variable already has a type attribute:
> https://github.com/RobotLocomotion/drake/blob/master/drake/solvers/decision_variable.h#L22
> and I think it's more correct for the variable to own it's type than for the mathematicalprogram to own it.  can we zap this variable and use the other data instead?
</details>
We do not have a method in `MathematicalProgram` to get the decision variables,  or a method to get the type from `DecisionVariable`.  I create a getter for the type from `DecisionVariable`, and a function `VariableTypes` to get a vector containing the type of each variable in `MathematicalProgram`, by looping through each variable. Now `MathematicalProgram` does not own the type.

---

*[drake/solvers/mosek_solver.cc, line 334 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCnJ031jh0bllubrZd:-KVCqaig3_L6OpmdFvQu:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1b2a663430020208a75e17bb8a05247d5f4c4e9e/drake/solvers/mosek_solver.cc#L334)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, consider initializing these to NaN in case the method below doesn't actually set them for some reason (an error or bug in Mosek?).
</details>
Done.

---

*[drake/solvers/mosek_solver.cc, line 354 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCngnx4yfEsT2USVLD:-KVCqb7G3_kj2kGlMgAK:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1b2a663430020208a75e17bb8a05247d5f4c4e9e/drake/solvers/mosek_solver.cc#L354)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, should be "// namespace *whatever*"
</details>
Done.

---

*[drake/solvers/mosek_solver.cc, line 391 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCjD1TZZ_0lqv6naiB:-KVCqgQ_GvQluAPzYWBd:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1b2a663430020208a75e17bb8a05247d5f4c4e9e/drake/solvers/mosek_solver.cc#L391)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> typo
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-:-KVCweCgfCM_bqp2-YmN:bny4ein)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[drake/solvers/mathematical_program.h, line 694 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCxbp9qHH_UFvklAPs:-KVCxbp9qHH_UFvklAPt:b-lcvymy) ([raw file](https://github.com/robotlocomotion/drake/blob/1cd93931c6daefb67a7b84f188787cd419e35c66/drake/solvers/mathematical_program.h#L694)):*
> ```objc
>    * The code above imposes constraints
>    * \f[
>    * -x(2) + 2x(5) = 1 \\
> ```

have to use \newline  instead
https://github.com/RobotLocomotion/drake/blob/master/drake/examples/simple_continuous_time_system.cc#L16

---

*[drake/solvers/mathematical_program.h, line 1239 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCy5jKQ9tcBr520Xbg:-KVCy5jLedwxkaQupLIC:b-rq5txo) ([raw file](https://github.com/robotlocomotion/drake/blob/1cd93931c6daefb67a7b84f188787cd419e35c66/drake/solvers/mathematical_program.h#L1239)):*
> ```objc
>    * decision variables.
>    */
>   void VariableTypes(
> ```

is this really better than looping through the variables in the solver interface?
and if so (for encapsulation), why not just allocate and return the std::vector?

that would also be much more consistent with the rest of the method signatures in this class.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3948)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 7 of 10 files reviewed at latest revision, 15 unresolved discussions.

---

*[drake/solvers/mathematical_program.h, line 381 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCkWMx3tTHsasO_0H6:-KVCy1tIoBNsnGfwhgXh:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1b2a663430020208a75e17bb8a05247d5f4c4e9e/drake/solvers/mathematical_program.h#L381)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Does it make sense to have a default name here? If I add two binary variables taking the default is it OK that they are both named "b"? Or does it make "b1" and "b2"? Comment should clarify.
</details>
Done.

---

*[drake/solvers/mathematical_program.h, line 381 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCjxedMsBTiApkvuHa:-KVD3wumS1_1X78-wzzB:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1b2a663430020208a75e17bb8a05247d5f4c4e9e/drake/solvers/mathematical_program.h#L381)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW: reflow to get rid of nearly-blank line above; capitalize "number"; not sure `@default` is what you want here -- did you generate the doxygen and look at it?
</details>
Done.

---

*[drake/solvers/mathematical_program.h, line 690 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVClv0IM5jW6ahbfGWm:-KVCzk3WqQ4b3l7ag5oE:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1b2a663430020208a75e17bb8a05247d5f4c4e9e/drake/solvers/mathematical_program.h#L690)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, typo and->an
</details>
Done.

---

*[drake/solvers/mathematical_program.h, line 692 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVClKLekj9u6pmnIvCk:-KVCzugotzRUunIkLFQS:b-xvsdm7) ([raw file](https://github.com/robotlocomotion/drake/blob/1b2a663430020208a75e17bb8a05247d5f4c4e9e/drake/solvers/mathematical_program.h#L692)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, use `@code`; we don't allow doxygen backslashes. Also, please verify that doxygen recognizes ".cc" as C++. (Generate and take a look if you haven't already.)
</details>
Done. 

I am pretty sure I compiled the documentation on my machine by 
```
$ cd drake-distro/build/drake/doc
$ make documentation
```
And doxygen recognizes the code block with backslashes. Why we don't allow backslashes?

I confirm @code works and recognizes ".cc" as C++.

---

*[drake/solvers/mathematical_program.h, line 694 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCxbp9qHH_UFvklAPs:-KVD0CtK-AKVmgbvkq4g:b4gb18h) ([raw file](https://github.com/robotlocomotion/drake/blob/1cd93931c6daefb67a7b84f188787cd419e35c66/drake/solvers/mathematical_program.h#L694)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> have to use \newline  instead
> https://github.com/RobotLocomotion/drake/blob/master/drake/examples/simple_continuous_time_system.cc#L16
</details>
![Screenshot from 2016-10-28 21:26:37.png](https://files.reviewableusercontent.io/images/reviewable/6314000/vAcW7MTzr7w1Ni9nKYW2PZiY/Screenshot%20from%202016-10-28%2021%3A26%3A37.png)
The picture above is the doxygen generated using `\f[, \f]` and `\\` for new line.

I thought by using `\f[` and `\f]`, I can use `\\` to indicate new line, at least it works on my local machine. Are we forbidding using `\f[` and `\f]`? It works like a multi-line math environment in latex. https://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmdfbropen

I am not sure of the doxygen in https://github.com/RobotLocomotion/drake/blob/master/drake/examples/simple_continuous_time_system.cc#L16, since it is in a .cc file, not in the public API. I believe doxygen is not generated for this comment.

When I tried to use `\begin{gather}` and `\end{gather}` on my machine, doxygen does not parse it correctly.

---

*[drake/solvers/mathematical_program.h, line 1239 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCy5jKQ9tcBr520Xbg:-KVD2ejFvs3JquoEQvth:bplefhc) ([raw file](https://github.com/robotlocomotion/drake/blob/1cd93931c6daefb67a7b84f188787cd419e35c66/drake/solvers/mathematical_program.h#L1239)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> is this really better than looping through the variables in the solver interface?
> and if so (for encapsulation), why not just allocate and return the std::vector?
> 
> that would also be much more consistent with the rest of the method signatures in this class.
</details>
The solver cannot access the variables stored inside `MathematicalProgram`, so it cannot loop through it (I think). And this function can be used by all solvers, so each solver should implement their own function to loop through the variables.

I think you are right on returning the std::vector. I changed it already.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3948)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 7 of 10 files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/solvers/mathematical_program.h, line 1239 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCy5jKQ9tcBr520Xbg:-KVD8Alm0w3Fk8Wrufz-:b-q6cn4u) ([raw file](https://github.com/robotlocomotion/drake/blob/1cd93931c6daefb67a7b84f188787cd419e35c66/drake/solvers/mathematical_program.h#L1239)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> The solver cannot access the variables stored inside `MathematicalProgram`, so it cannot loop through it (I think). And this function can be used by all solvers, so each solver should implement their own function to loop through the variables.
> 
> I think you are right on returning the std::vector. I changed it already.
</details>
Sorry, I mean each solver should *not* loop through the variables, but rather the common `MathematicalProgram` should do.

I am a bit confused. Do you think we should expose the variables stored in `MathematicalProgram` to the solver interface? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3948)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[drake/solvers/gurobi_solver.cc, line 382 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVD8uqbFT97PKYDi1fB:-KVD8uqcLsn0nx8Xs1Cn:b-om8tc5) ([raw file](https://github.com/robotlocomotion/drake/blob/70470899ef951a9685b73acff4a10a5b59e66ad5/drake/solvers/gurobi_solver.cc#L382)):*
> ```cpp
>   std::vector<double> xupp(num_vars, std::numeric_limits<double>::infinity());
> 
>   const std::vector<DecisionVariable::VarType>& var_type = prog.VariableTypes();
> ```

sorry if I'm being slow, but how can this be a reference?
(I would have used auto var_type = )

---

*[drake/solvers/mathematical_program.h, line 694 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCxbp9qHH_UFvklAPs:-KVD87O4YHJzZBDUYB8N:b-11byr6) ([raw file](https://github.com/robotlocomotion/drake/blob/1cd93931c6daefb67a7b84f188787cd419e35c66/drake/solvers/mathematical_program.h#L694)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> ![Screenshot from 2016-10-28 21:26:37.png](https://files.reviewableusercontent.io/images/reviewable/6314000/vAcW7MTzr7w1Ni9nKYW2PZiY/Screenshot%20from%202016-10-28%2021%3A26%3A37.png)
> The picture above is the doxygen generated using `\f[, \f]` and `\\` for new line.
> 
> I thought by using `\f[` and `\f]`, I can use `\\` to indicate new line, at least it works on my local machine. Are we forbidding using `\f[` and `\f]`? It works like a multi-line math environment in latex. https://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmdfbropen
> 
> I am not sure of the doxygen in https://github.com/RobotLocomotion/drake/blob/master/drake/examples/simple_continuous_time_system.cc#L16, since it is in a .cc file, not in the public API. I believe doxygen is not generated for this comment.
> 
> When I tried to use `\begin{gather}` and `\end{gather}` on my machine, doxygen does not parse it correctly.
</details>
i think the \gather in the other file is a defect.  (if you're willing, you could change those to \f for me?).  but replacing \\ with \newline was to address something that @jwnimmer-tri hit on a different compiler/platform.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3948)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 10 of 11 files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/solvers/gurobi_solver.cc, line 382 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVD8uqbFT97PKYDi1fB:-KVDATdHc-NujkhC9Ute:b-ozsp2r) ([raw file](https://github.com/robotlocomotion/drake/blob/70470899ef951a9685b73acff4a10a5b59e66ad5/drake/solvers/gurobi_solver.cc#L382)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> sorry if I'm being slow, but how can this be a reference?
> (I would have used auto var_type = )
> 
</details>
I think `prog.variable_type()` returns a temporary, and the lifetime of this temporary object is bound to the reference `var_type`. It is explained here
https://herbsutter.com/2008/01/01/gotw-88-a-candidate-for-the-most-important-const/

---

*[drake/solvers/mathematical_program.h, line 694 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCxbp9qHH_UFvklAPs:-KVDBwFBlXHAAHTOHANn:b-490r4t) ([raw file](https://github.com/robotlocomotion/drake/blob/1cd93931c6daefb67a7b84f188787cd419e35c66/drake/solvers/mathematical_program.h#L694)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> i think the \gather in the other file is a defect.  (if you're willing, you could change those to \f for me?).  but replacing \\ with \newline was to address something that @jwnimmer-tri hit on a different compiler/platform.
</details>
Done. Ah sorry I did not know there is a compiler/platform issue with doxygen.

But when I use `\newline` in `\f[` and `\f]` environment, doxygen is not happy. Below is the screenshot of the documentation on my computer. I am using Ubuntu 14.04, the default compiler is gcc-4.7.
![Screenshot from 2016-10-28 22:14:10.png](https://files.reviewableusercontent.io/images/reviewable/6314000/oMzxHDxw7zsJu4wQ7hQ6_VPz/Screenshot%20from%202016-10-28%2022%3A14%3A10.png)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3948)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[drake/examples/simple_continuous_time_system.cc, line 16 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVDF0cemMAGop2YTWVf:-KVDF0cemMAGop2YTWVg:b-2tcn3k) ([raw file](https://github.com/robotlocomotion/drake/blob/25c51ff4ffc48c568764950d4a98b9e43a71ff90/drake/examples/simple_continuous_time_system.cc#L16)):*
> ```cpp
> /// Simple Continuous Time System
> /// \f[
> ///     \dot{x} = -x + x^3 \\
> ```

no, sorry.  i believe it should be \f[ (probably @f is better) but use \newline.
or there is a working example here:
https://github.com/RobotLocomotion/drake/blob/de3a220668f05662eea1cc4ee3abf3124d7b4d5a/drake/systems/framework/primitives/affine_system.h#L13

---

*[drake/solvers/gurobi_solver.cc, line 382 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVD8uqbFT97PKYDi1fB:-KVDEuHGpD4sHeZXez5k:b-3t7j) ([raw file](https://github.com/robotlocomotion/drake/blob/70470899ef951a9685b73acff4a10a5b59e66ad5/drake/solvers/gurobi_solver.cc#L382)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> I think `prog.variable_type()` returns a temporary, and the lifetime of this temporary object is bound to the reference `var_type`. It is explained here
> https://herbsutter.com/2008/01/01/gotw-88-a-candidate-for-the-most-important-const/
</details>
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> super useful.  thanks.

---

*[drake/solvers/mathematical_program.h, line 694 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCxbp9qHH_UFvklAPs:-KVDG6yJ1PZGtZTEeeF2:b-k7r5n6) ([raw file](https://github.com/robotlocomotion/drake/blob/1cd93931c6daefb67a7b84f188787cd419e35c66/drake/solvers/mathematical_program.h#L694)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> Done. Ah sorry I did not know there is a compiler/platform issue with doxygen.
> 
> But when I use `\newline` in `\f[` and `\f]` environment, doxygen is not happy. Below is the screenshot of the documentation on my computer. I am using Ubuntu 14.04, the default compiler is gcc-4.7.
> ![Screenshot from 2016-10-28 22:14:10.png](https://files.reviewableusercontent.io/images/reviewable/6314000/oMzxHDxw7zsJu4wQ7hQ6_VPz/Screenshot%20from%202016-10-28%2022%3A14%3A10.png)
> 
</details>
then probably follow the affine example linked above instead.  we should clearly document our mathjax usage.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3948)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 9 of 11 files reviewed at latest revision, 13 unresolved discussions.

---

*[drake/examples/simple_continuous_time_system.cc, line 16 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVDF0cemMAGop2YTWVf:-KVDIiobc8PD9wOsK7gq:bs7wumx) ([raw file](https://github.com/robotlocomotion/drake/blob/25c51ff4ffc48c568764950d4a98b9e43a71ff90/drake/examples/simple_continuous_time_system.cc#L16)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> no, sorry.  i believe it should be \f[ (probably @f is better) but use \newline.
> or there is a working example here:
> https://github.com/RobotLocomotion/drake/blob/de3a220668f05662eea1cc4ee3abf3124d7b4d5a/drake/systems/framework/primitives/affine_system.h#L13
> 
</details>
Done. 

---

*[drake/solvers/mathematical_program.h, line 694 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVCxbp9qHH_UFvklAPs:-KVDI_nfD-yT7x_Hp57f:b2h6sbp) ([raw file](https://github.com/robotlocomotion/drake/blob/1cd93931c6daefb67a7b84f188787cd419e35c66/drake/solvers/mathematical_program.h#L694)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> then probably follow the affine example linked above instead.  we should clearly document our mathjax usage.
</details>
Done. I agree we should document our mathjax usage, and make sure this documentation works across platform/compilers.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3948)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-:-KVDJICZVGTv2tiJTCIx:bnfp4nl)*
<!-- Sent from Reviewable.io -->

rdeits commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 5 files at r2, 1 of 4 files at r3, 2 of 3 files at r4, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-:-KVDMj7MeBcvlp_6o8xD:bnfp4nl)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> - all my comments are unblocked.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/solvers/mathematical_program.h, line 692 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVClKLekj9u6pmnIvCk:-KVGkPr9NkEhDjyhfz1F:bgaf4q8) ([raw file](https://github.com/robotlocomotion/drake/blob/1b2a663430020208a75e17bb8a05247d5f4c4e9e/drake/solvers/mathematical_program.h#L692)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> Done. 
> 
> I am pretty sure I compiled the documentation on my machine by 
> ```
> $ cd drake-distro/build/drake/doc
> $ make documentation
> ```
> And doxygen recognizes the code block with backslashes. Why we don't allow backslashes?
> 
> I confirm @code works and recognizes ".cc" as C++.
</details>
We decided that the backslashes were dangerous in C++ code where backslashes already have a lot of significance. Since the `@`s work we want to use them exclusively.

---

*[drake/solvers/mathematical_program.h, line 412 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVGk9BlTU8_1PXXCezF:-KVGk9BlTU8_1PXXCezG:b50ir84) ([raw file](https://github.com/robotlocomotion/drake/blob/aa36186802b949fe869b4d377b9d211fb8de1036/drake/solvers/mathematical_program.h#L412)):*
> ```objc
>    * auto b2 = prog.AddBinaryVariables(2, "b");
>    * @endcode
>    * The name of the variable is only used for user for understand the problem.
> ```

BTW, reword: "only used for the user to understand"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-:-KVGkrMgkKhbW_jtQj5Q:b-bcvts0)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 10 of 11 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/mathematical_program.h, line 412 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVGk9BlTU8_1PXXCezF:-KVHrGvpcwleANAGqbY1:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/aa36186802b949fe869b4d377b9d211fb8de1036/drake/solvers/mathematical_program.h#L412)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, reword: "only used for the user to understand"
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3948)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
@sherm1 , since I hit the "Done" button, you will need to acknowledge my change again. Sorry, I will use "Acknowledge" button next time.

---

Review status: 10 of 11 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-:-KVHr_JIqgSTyCiiIaEr:b-puz7y3)*
<!-- Sent from Reviewable.io -->

liangfok commented:
BTW, since @sherm1's comment was labeled "BTW", and you have obviously accepted and applied his suggestion, I think it'll be OK for you to dismiss him from the conversation so this PR can be merged.

---

Review status: 10 of 11 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-:-KVHwe5KlP4kTngoZVYh:b-9dv6qb)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 10 of 11 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/mathematical_program.h, line 412 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVGk9BlTU8_1PXXCezF:-KVHx9PD_BfJASrmRcjO:bq9qum3) ([raw file](https://github.com/robotlocomotion/drake/blob/aa36186802b949fe869b4d377b9d211fb8de1036/drake/solvers/mathematical_program.h#L412)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> Done.
</details>
BTW, I notice there is a "the" missing before "user".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3948)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 10 of 11 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/mathematical_program.h, line 412 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/3948#-KVGk9BlTU8_1PXXCezF:-KVI3RdmXEGK4TUE7LB3:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/aa36186802b949fe869b4d377b9d211fb8de1036/drake/solvers/mathematical_program.h#L412)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW, I notice there is a "the" missing before "user".
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3948)*
<!-- Sent from Reviewable.io -->
